### PR TITLE
Homa nominate validators

### PR DESCRIPTION
**Pull request by wangjj9219**

### CODE REVIEW

File: modules/homa-validator-list/src/lib.rs [hunk 1]
```diff
@@ -505,13 +500,70 @@ pub mod module {
 			T::OnSlash::happened(&actual_total_slashing);
 			Ok(())
 		}
+
+		#[pallet::call_index(7)]
+		#[pallet::weight(T::WeightInfo::set_reserved_validators(updates.len() as u32))]
+		pub fn set_reserved_validators(
+			origin: OriginFor<T>,
+			updates: Vec<(u16, Vec<T::RelayChainAccountId>)>,
```
[the remark this answers is not in the document]
Comment (wangjj9219) [reply modules/homa-validator-list/src/lib.rs:508]: param`updates` will be checked and converted  to `BoundedVec` type in the function body before insert storage .

[the remark this answers is not in the document]
Comment (xlc) [reply modules/homa-validator-list/src/lib.rs:508]: but it is still better to validate it at the time of construct the parameter so a transaction with too many nominations won't even be able to make into tx pool

File: runtime/karura/src/lib.rs [hunk 2]
```diff
@@ -1579,6 +1579,32 @@ impl module_homa::Config for Runtime {
...
+	type RuntimeEvent = RuntimeEvent;
+	type RelayChainAccountId = AccountId;
+	type LiquidTokenCurrency = module_currencies::Currency<Runtime, GetLiquidCurrencyId>;
+	type MinBondAmount = MinBondAmount;
+	type BondingDuration = ValidatorBackingBondingDuration;
+	type ValidatorInsuranceThreshold = ValidatorInsuranceThreshold;
+	type GovernanceOrigin = EnsureRootOrHalfGeneralCouncil;
+	type OnSlash = ();
+	type LiquidStakingExchangeRateProvider = Homa;
+	type WeightInfo = ();
```
Comment (zjb0807) [inline runtime/karura/src/lib.rs:1602]: update the WeightInfo

**Review by zjb0807 — APPROVED** [review]

File: runtime/acala/src/lib.rs [hunk 3]
```diff
@@ -1554,6 +1554,32 @@ impl module_homa::Config for Runtime {
...
 	type XcmInterface = XcmInterface;
 	type WeightInfo = weights::module_homa::WeightInfo<Runtime>;
+	type NominationsProvider = HomaValidatorList;
+}
+
+parameter_types! {
+	pub MinBondAmount: Balance = 10 * dollar(LDOT);
+	pub const ValidatorBackingBondingDuration: BlockNumber = 28 * DAYS;
+	pub ValidatorInsuranceThreshold: Balance = 100_000 * dollar(LDOT);
+	pub const MaxNominations: u32 = 24;
```
Comment (xlc) [inline runtime/acala/src/lib.rs:1564]:
```suggestion
	pub const MaxNominations: u32 = 16;
```
only 16 for polkadot

File: runtime/acala/src/lib.rs [hunk 4]
```diff
@@ -1554,6 +1554,32 @@ impl module_homa::Config for Runtime {
 	type RelayChainBlockNumber = RelaychainDataProvider<Runtime>;
 	type XcmInterface = XcmInterface;
 	type WeightInfo = weights::module_homa::WeightInfo<Runtime>;
+	type NominationsProvider = HomaValidatorList;
```
Comment (xlc) [inline runtime/acala/src/lib.rs:1557]: this should be reading from nominees-election pallet

File: runtime/acala/src/lib.rs [hunk 5]
```diff
@@ -1554,6 +1554,32 @@ impl module_homa::Config for Runtime {
...
+}
+
+parameter_types! {
+	pub MinBondAmount: Balance = 10 * dollar(LDOT);
```
Comment (xlc) [inline runtime/acala/src/lib.rs:1561]:
```suggestion
	pub MinBondAmount: Balance = 1000 * dollar(LDOT);
```

File: runtime/acala/src/lib.rs [hunk 6]
```diff
@@ -1554,6 +1554,32 @@ impl module_homa::Config for Runtime {
...
+	pub const ValidatorBackingBondingDuration: BlockNumber = 28 * DAYS;
+	pub ValidatorInsuranceThreshold: Balance = 100_000 * dollar(LDOT);
```
Comment (xlc) [inline runtime/acala/src/lib.rs:1563]:
```suggestion
	pub ValidatorInsuranceThreshold: Balance = 10_000 * dollar(LDOT);
```

**Review by xlc — CHANGES_REQUESTED** [review]
should also integrate with nominees-election to allow LDOT holders to elect validators

**Comment by wangjj9219** [comment]
> should also integrate with nominees-election to allow LDOT holders to elect validators

reviewed the designs of `nominee-election` and `homa-validator-list`. They are similar. It seems that `nominee-election` is an earlier design. Later, there was the pallet `homa-validator-list`. Similarly, LDOT holders can give validators Votes to increase its guarantee, and the locked guarantee can be frozen and slashed through governance

Need to confirm, is the design of `homa-validator-list` enough to replace `nominee-election`? Does it need optimization, or some renaming?

**Comment by xlc** [comment]
This is the intended usage:

Validators what wish to become LDOT validators needs to register themselves using `homa-validator-list` and provide insurance deposit.

Then LDOT holders can nominate LDOT validators using `nominee-election` to pick the top X validators out of all registered validators from `homa-validator-list`

So the deposit for `homa-validator-list` can be low as they have to be top X picked by LDOT holders anyway

**Comment by wangjj9219** [comment]
> This is the intended usage:
> 
> Validators what wish to become LDOT validators needs to register themselves using `homa-validator-list` and provide insurance deposit.
> 
> Then LDOT holders can nominate LDOT validators using `nominee-election` to pick the top X validators out of all registered validators from `homa-validator-list`
> 
> So the deposit for `homa-validator-list` can be low as they have to be top X picked by LDOT holders anyway

I understand, `homa-validator-list` serves as a filter, and the pick up logic should be placed in `nominee-election`, and `homa` read nominee from `nominee-election`.

File: runtime/karura/src/lib.rs [hunk 7]
```diff
@@ -1573,12 +1575,53 @@ impl module_homa::Config for Runtime {
...
 	type RedeemThreshold = RedeemThreshold;
 	type RelayChainBlockNumber = RelaychainDataProvider<Runtime>;
 	type XcmInterface = XcmInterface;
 	type WeightInfo = weights::module_homa::WeightInfo<Runtime>;
+	type NominationsProvider = NomineesElection;
+}
+
+parameter_types! {
+	pub MinBondAmount: Balance = 10 * dollar(LKSM);
+	pub ValidatorInsuranceThreshold: Balance = 10_000 * dollar(LKSM);
```
Comment (xlc) [inline runtime/karura/src/lib.rs:1589]:
```suggestion
	pub MinBondAmount: Balance = 100 * dollar(LKSM);
	pub ValidatorInsuranceThreshold: Balance = 1_000 * dollar(LKSM);
```

File: runtime/acala/src/lib.rs [hunk 8]
```diff
@@ -1548,12 +1550,53 @@ impl module_homa::Config for Runtime {
...
+	type BondingDuration = BondingDuration;
+	type ValidatorInsuranceThreshold = ValidatorInsuranceThreshold;
+	type GovernanceOrigin = EnsureRootOrHalfGeneralCouncil;
+	type LiquidStakingExchangeRateProvider = Homa;
+	type CurrentEra = Homa;
+	type WeightInfo = weights::module_homa_validator_list::WeightInfo<Runtime>;
+}
+
+parameter_types! {
+	pub MinCouncilBondThreshold: Balance = dollar(LDOT);
```
Comment (xlc) [inline runtime/acala/src/lib.rs:1581]:
```suggestion
	pub MinNomineesElectionBondThreshold: Balance = 10 * dollar(LDOT);
```

File: runtime/karura/src/lib.rs [hunk 9]
```diff
@@ -1573,12 +1575,53 @@ impl module_homa::Config for Runtime {
...
+	type BondingDuration = BondingDuration;
+	type ValidatorInsuranceThreshold = ValidatorInsuranceThreshold;
+	type GovernanceOrigin = EnsureRootOrHalfGeneralCouncil;
+	type LiquidStakingExchangeRateProvider = Homa;
+	type CurrentEra = Homa;
+	type WeightInfo = weights::module_homa_validator_list::WeightInfo<Runtime>;
+}
+
+parameter_types! {
+	pub MinCouncilBondThreshold: Balance = dollar(LKSM);
```
Comment (xlc) [inline runtime/karura/src/lib.rs:1606]:
```suggestion
	pub MinNomineesElectionBondThreshold: Balance = dollar(LKSM);
```

**Review by xlc — DISMISSED** [review]

**Review by xlc — APPROVED** [review]

**Pull request merged**